src/IssueViz.Security/Taint/TaintList/ViewModels/TaintIssuesControlViewModel.cs

```diff
@@ -0,0 +1,120 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Collections.Specialized;
+using System.Linq;
+using System.Windows.Data;
+using System.Windows.Input;
+using Microsoft.VisualStudio.PlatformUI;
+using Microsoft.VisualStudio.Shell;
+using SonarLint.VisualStudio.IssueVisualization.Editor;
+using SonarLint.VisualStudio.IssueVisualization.Models;
+
+namespace SonarLint.VisualStudio.IssueVisualization.Security.Taint.TaintList.ViewModels
+{
+    internal interface ITaintIssuesControlViewModel : IDisposable
+    {
+        // WIP: add selection handling
+        // WIP: add any other necessary commands
+
+        ICommand NavigateCommand { get; }
+
+        ObservableCollection<ITaintIssueViewModel> Issues { get; }
+    }
+
+    /// <summary>
+    /// View model that surfaces the data in the TaintStore
+    /// </summary>
+    /// <remarks>Monitors the taint store for changes and updates the view model accordingly.</remarks>
+    internal sealed class TaintIssuesControlViewModel : ITaintIssuesControlViewModel
```

Comment:
Looking at the code, I think the sheer amount of duplication is worthy of a generic component.

I was thinking that maybe we can do something like this:
```
IIssueViewModel
{
 	IAnalysisIssueVisualization TaintIssueViz { get; }
        int Line { get; }
        int Column { get; }
        string DisplayPath { get; }
}

IIssuesControlViewModel<T> where T : IIssueViewModel
{
	ICommand NavigateCommand { get; }
        ObservableCollection<T> Issues { get; }
}

IHotspotViewModel : IIssueViewModel
{
	string CategoryDisplayName { get; }
}

IHotspotsControlViewModel : IIssuesControlViewModel<IHotspotViewModel>
{
	IHotspotViewModel SelectedHotspot { get; }
	ICommand RemoveCommand { get; }
}
```

And create a base class that can contain essentially everything that is in TaintIssuesControlViewModel.
WDYT?

Reply:
Yes, I think there is too much duplication and there should be a couple of view model base classes (that's what I was alluding to when I mentioned identifying opportunities for refactoring above - perhaps I should have been more explicit). My plan was to make the changes to the taint tool window and control to get a working end-to-end where both view models are actually being consumed and then refactor to extract the common behaviour.

Reply:
FYI one part of the potentially common view model behaviour that I'm not currently sure about is selection, more specifically integration with the selection service. We currently have two selection services, `IAnalysisIssueSelectionService` and `IHotspotSelectionService`. I don't think we want to add a third one for taint. We might end up with one service that can be consumed in a consistent way by the editor light bulbs, flow window and error/hotspot/taint lists.